### DISCUSSION TITLE
Serializing and Deserializing Expression Trees to and from json

### DISCUSSION THREAD

**Discussion by chrisserrao** [dotnet/csharplang#5555] [Language Ideas]
This is my first post, so please excuse any mistakes I make.

Can we have the ability to serialize and deserialize expression trees.

**Scenario:**
I have a piece of functionality where the logic that does the calculation or validation can change frequently or depending on our customer.
I would like to create an expression lambda (a block expression) which I can serialize and save in the database and then the code can read and deserialize the expression into an expression tree, which I can then run on production.

**What we've tried:**
To achieve something similar I've tried CSharpScript, however, it was reported that the introduction of CSharpScript caused more memory to be used, not to mention more time consumed to compile and run the code.
I believe expression trees should be faster than CSharpScript and also they should consume less memory to cache a compiled expression lambda in memory so that it can be reused more easily.

Please let me know if this seems reasonable.
Looking forward to your expert comments and feedback.

**Comment by HaloFour** [comment]
This sounds like more of an ask of the [runtime repository](https://github.com/dotnet/runtime/), which handles the APIs and base class libraries and manages the classes used for expression trees.  However, I would think that with any JSON library that supports custom serialization/deserialization that functionality like this could be achieved entirely without language or runtime support, e.g. by implementing a [`JsonConverter<T>`](https://www.newtonsoft.com/json/help/html/CustomJsonConverterGeneric.htm) for JSON.NET.

**Reply by chrisserrao** [reply]
I actually did try using JsonConvert to serialize the expression but I kept running into a self referencing loop error (I think that was what it was called) when I tried to do string concatenation in my expression tree.

**Comment by YairHalberstadt** [comment]
If efficiency is important, how about writing ordinary c# code, compiling it, and storing the DLL in the database as a binary blob? Then you can load the DLL and run it

**Comment by TahirAhmadov** [comment]
The LINQ expression trees are not serialization friendly out of the box, regardless of what JSON library you use. You would have to write a custom serializer/deserializer (or perhaps somebody else created a LINQ expression tree serializer on Nuget?)

**Reply by svick** [reply]
> or perhaps somebody else created a LINQ expression tree serializer on Nuget?

[Somebody did.](https://github.com/6bee/Remote.Linq)

The following code works for me:

```c#
JsonSerializerOptions serializerSettings = new JsonSerializerOptions().ConfigureRemoteLinq();

Expression<Func<int, string>> e = (int i) => "The number is: " + i;
string json = JsonSerializer.Serialize(e.ToRemoteLinqExpression(), serializerSettings);
var deserialized = (Expression<Func<int, string>>)JsonSerializer.Deserialize<Remote.Linq.Expressions.LambdaExpression>(json, serializerSettings).ToLinqExpression();
```

The produced JSON is:

<details><summary>JSON</summary>

```json
{
  "$id": "1",
  "$type": "LambdaExpression",
  "Expression": {
    "$id": "2",
    "$type": "BinaryExpression",
    "BinaryOperator": 0,
    "LeftOperand": {
      "$id": "3",
      "$type": "ConstantExpression",
      "Type": "string",
      "Value": "The number is: "
    },
    "RightOperand": {
      "$id": "4",
      "$type": "UnaryExpression",
      "UnaryOperator": 10,
      "Operand": {
        "$id": "5",
        "$type": "ParameterExpression",
        "ParameterName": "i",
        "ParameterType": "int32",
        "InstanceId": 1
      },
      "Type": {
        "$id": "6",
        "$type": "type",
        "Name": "Object",
        "Namespace": "System"
      }
    },
    "Method": {
      "$id": "7",
      "$type": "MethodInfo",
      "Name": "Concat",
      "DeclaringType": "string",
      "IsStatic": true,
      "ParameterTypes": {
        "$id": "8",
        "$values": [
          {
            "$ref": "6"
          },
          {
            "$ref": "6"
          }
        ]
      },
      "ReturnType": "string"
    }
  },
  "Parameters": {
    "$id": "9",
    "$values": [
      {
        "$ref": "5"
      }
    ]
  },
  "Type": {
    "$id": "10",
    "$type": "type",
    "Name": "Func\u00602",
    "Namespace": "System",
    "GenericArguments": {
      "$id": "11",
      "$values": [
        "int32",
        "string"
      ]
    },
    "IsGenericType": true
  }
}
```

</details>

**Reply by chrisserrao** [reply]
Thanks, this is great. However, I think one challenge is to create block expressions in code. I tried searching, but I think the only way to create block expressions is by programmatically creating them using the Expressions API. It would be great if we could create them directly in code.

**Reply by svick** [reply]
@chrisserrao That's https://github.com/dotnet/csharplang/discussions/158.

**Reply by chrisserrao** [reply]
Thank you so much !

**Reply by bartdesmet** [reply]
You can also try https://github.com/reaqtive/reaqtor/tree/main/Nuqleon/Core/LINQ/Nuqleon.Linq.Expressions.Bonsai.Serialization. It's been used in production at Microsoft for many years to power distributed event processing systems etc. There's a corresponding NuGet package, and there's also a .NET Interactive notebook containing sample usage code in the folder linked above. It supports all nodes in .NET expression tree APIs (including "statement" nodes).

**Reply by IbrahimElshafey** [reply]
@bartdesmet Thanks a lot for your comment.
